UI: Repo layer integration tests

johncowen opened the pull request:
This PR mainly adds a set of integration tests to the highest level of my model/data layer, basically just above my 'repo' Services.

The aim being, now that the Datacenters are added to the ember-data object pre-serialization instead of post-serialization (see #4398 ), to enable me to confidently move my data serializing out of the adapters and into the serializers where people expect them to be.

It also adds the beginning of being able to test performance of larger datasets moving forwards via a configurable `measure` function, that simply console.log's the results out unless you add a custom reporter to output to something external.

In putting these together it also surfaced a couple of trivial 'bugs' that aren't affecting anything, but I've fixed them here anyway before I go about with the adapter > serializer migration.

The `repo` function itself goes through a number of steps in order to retrieve a set of mock data, execute and measure the test on its own and then provide an entrypoint to allow you to mutate the mock data to how it should be and assert your results. I've commented it up a little as but hopefully you can see that how its made writing the tests easier.

I've also been toying with ideas for eliminating the need to fill out the `needs` array, I'm thinking this should automatically know about all its dependencies.

I'm PR'ing this now in order to keep things easier to digest. Once I get approval on this I can go about moving everything about.

johncowen commented on ui-v2/app/models/dc.js:
```diff
@@ -9,6 +9,7 @@ export const SLUG_KEY = 'Name';
+  // TODO: Are these required?
   Services: hasMany('service'),
   Nodes: hasMany('node'),
 });
```
This is just a little note for later, right now these aren't being used.

johncowen commented on ui-v2/app/services/session.js:
```diff
@@ -9,6 +9,7 @@ export default Service.extend({
+  // TODO: Why Key? Probably should be findBySlug like the others
   findByKey: function(slug, dc) {
     return get(this, 'store').queryRecord('session', {
       id: slug,
```
Again a little note for later. Some of the names of these don't quite follow the same naming system, I'm sure there was a reason for that when I did it. But once this is all complete I'll probably review the names and make sure they follow the same naming pattern - and if they don't - that they have a reason for it.

johncowen commented on ui-v2/tests/helpers/api.js:
```diff
@@ -27,6 +27,7 @@ export const get = function(_url, options = { headers: { cookie: {} } }) {
       },
       {
+        set: function() {},
         send: function(content) {
```
This looks weird I know! 😄

I noticed that when my api-double tries to send headers when its being run as a HTTP server it calls this function, everything is fine when its running over HTTP, but when I'm running the double in the HTML page without the server it looks to call this if there are headers to be sent. Right now I have no tests that are testing what comes back from consul, and I not sure if I ever will, so for now at least this is a noop when used in the frontend/without a server.

johncowen commented on ui-v2/tests/helpers/measure.js:
```diff
@@ -0,0 +1,5 @@
+import getMeasure from 'consul-ui/tests/lib/measure/getMeasure';
+let report;
+let len = 1;
+export default getMeasure(len, report);
```
`len` here is the amount of times to run a test in order to then take an average for results, when I switch in a different reporter I switch this up to a higher number, I could have commented this in a little more, but I'm imagining this is a first draft of the reporting stuff so it's subject to change. `report` is empty just because for the moment I'm just adding in a custom reporter here when I'm interested, likely to move this to something more 'inject-y' at a later date.

johncowen commented on ui-v2/tests/integration/services/dc-test.js:
```diff
@@ -0,0 +1,45 @@
+    'model:node',
+  ],
+});
+skip("findBySlug (doesn't interact with the API) but still needs an int test");
```
I'm skipping this one for now as this method doesn't interact with the API, but I'll get reminded to come back here at some point.

johncowen commented on ui-v2/tests/integration/services/services-test.js:
```diff
@@ -0,0 +1,75 @@
+    }
+  );
+});
+skip('findBySlug returns the correct data for item endpoint', function(assert) {
```
Ah I need to un skip this before I go about changing it - it was troublesome for some reason, will probably sort that tomorrow now, but would be good to get some eyes on this before I start changing things around.

johncowen commented on ui-v2/tests/lib/measure/getMeasure.js:
```diff
@@ -0,0 +1,26 @@
+        return actual;
+      });
+  };
+}
```
Here's the measurement bits, this one measures things that run a promise based API - I have another version of this on the sidelines for non-promise based API's just not sure how to have them automatically switch in and out yet, or even if I want it to do it automatically.

johncowen commented on ui-v2/tests/integration/services/services-test.js:
```diff
@@ -58,16 +59,22 @@ skip('findBySlug returns the correct data for item endpoint', function(assert) {
+              Datacenter: dc,
+              uid: `["${dc}","${id}"]`,
+            }
+          );
```
So this tree is not very nice, and wasn't what I was expecting. I've tested for it as this PR is to refactor what is there without introducing any regressions. I've added a note and a skipped test to remind me to come back here at a later date.

johncowen commented on ui-v2/tests/helpers/repo.js:
```diff
@@ -0,0 +1,96 @@
+import { get } from 'consul-ui/tests/helpers/api';
+import { get as _get } from '@ember/object';
```
Totally agree, fixed

johncowen commented on ui-v2/tests/helpers/repo.js:
```diff
@@ -0,0 +1,96 @@
+    });
+    tags = {
+      count: key.length > 0 ? parseInt(cookies[key[0]]) : 1,
+    };
```
Agree `find` is better here -- I had used `filter` as to my memory we were trying to support IE11, which I believe `find` is not compatible with (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find#Browser_compatibility). Do you know what our choice was here? Will definitely switch to find if there is no need for that.

DingoEatingFuzz replied:
I didn't realize IE11 didn't support it. Ember polyfills (or at least used to?) all these expected functions, so I've been using it for years.

johncowen commented on ui-v2/tests/helpers/repo.js:
```diff
@@ -0,0 +1,96 @@
+            return test(service);
+          },
+          `${name}Service.${method}`,
+          tags
```
Do you mind expanding a bit on what you mean?

DingoEatingFuzz replied:
The contents of tags changes in a couple places which can be surprising. Particularly because the line order doesn't equate to execution order and without a careful read a person may not realize the impact a refactor would have.

If instead were somehow computed at will, then they could be safely computed in both places they are needed and they would no longer be order-dependent.

johncowen commented on ui-v2/tests/helpers/repo.js:
```diff
@@ -0,0 +1,96 @@
+      adapter
+    );
+  });
+}
```
See below comment on the pseudo code example!

johncowen commented on ui-v2/tests/integration/services/acls-test.js:
```diff
@@ -0,0 +1,61 @@
+      );
+    }
+  );
+});
```
I agree your examples were clearer. It felt like a more functional approach was a better choice here given that is the precedent set in the rest of the project.

I thought adding function names might be a good compromise here? For example:

```
  return repo(
    'Service',
    'findAllByDatacenter',
    this.subject(),
    function retrieveStub(stub) {
      return stub(`/v1/internal/ui/services?dc=${dc}`, {
        CONSUL_SERVICE_COUNT: '100',
      });
    },
    function performTest(service) {
      return service.findAllByDatacenter(dc);
    },
    function performAssertion(actual, expected) {
      assert.deepEqual(
        actual,
        expected(function(payload) {
          return payload.map(item =>
            Object.assign({}, item, {
              Datacenter: dc,
              uid: `["${dc}","${item.Name}"]`,
            })
          );
        })
      );
    }
  );

```
..or similar names to help with self documentation?

Do you think this is a good compromise?

johncowen commented on ui-v2/tests/integration/services/coordinates-test.js:
```diff
@@ -0,0 +1,43 @@
+const NAME = 'coordinate';
+moduleFor(`service:${NAME}s`, `Integration | Service | ${NAME}s`, {
+  // Specify the other units that are required for this test.
+  needs: [
```
There is, but it is only needed for component integration tests to my understanding, and in this case I'm not testing a component so I don't think it is needed?

DingoEatingFuzz replied:
If I understand correctly, when you say `integration: true`, it wires up all the `needs` for you so you don't have to declare them explicitly. 

However, in a unit test, it's nice to be explicit with the dependencies and have tests fail when new dependencies are added without acknowledgement.

johncowen commented:
Hey @DingoEatingFuzz 

I've moved this to use `find` and added the function names throughout to help with readability so I'm going to go ahead and merge this now.

Thanks,
